src/meds_torch/models/components/transformer_decoder.py

```diff
@@ -32,3 +298,286 @@ def forward(self, batch):
+                0
+            ) < prompt_lengths.unsqueeze(1)
+            if not torch.equal(right_pad_mask, mask):
+                raise ValueError("Mask must correspond to right padding")
+
+        return generate(
+            self.model,
+            prompts=prompts,
+            budget=self.cfg.generation_budget,
+            get_next_token_time=get_next_token_time,
+            eos_token=eos_token,
+            temperature=temperature,
+            prompt_lens=prompt_lengths,
+            filter_logits_fn=filter_logits_fn,
+            restrict_to_max_seq_len=restrict_to_max_seq_len,
+            filter_kwargs=filter_kwargs,
+            cache_kv=cache_kv,
+            pad_value=pad_value,
+            **kwargs,
+        )
+
+
+@torch.no_grad()
+@eval_decorator
+def generate(
+    model: TransformerWrapper,
+    prompts: torch.Tensor,
+    budget: GenerationBudget,
+    get_next_token_time: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] | None = None,
+    eos_token=None,
+    temperature=1.0,
+    prompt_lens: torch.Tensor | None = None,
+    filter_logits_fn: str | Callable = identity,
+    restrict_to_max_seq_len: bool = True,
+    filter_kwargs: dict = dict(),
+    cache_kv=True,
+    pad_value: int = 0,
+    **kwargs,
+):
+    """Generate tokens autoregressively from given prompts.
+
+    Args:
+        prompts: Input token sequence [batch_size, seq_len]
+        seq_len: Number of tokens to generate
+        eos_token: Optional end of sequence token to stop generation
+        temperature: Sampling temperature (0 for greedy)
+        prompt_lens: Optional lengths for variable length prompts
+        filter_logits_fn: Optional filtering function for logits
+        restrict_to_max_seq_len: Whether to restrict to model's max sequence length
+        filter_kwargs: Additional arguments for filter_logits_fn
+        cache_kv: Whether to use key-value caching
+        **kwargs: Additional arguments passed to model forward
+
+    Returns:
+        Generated token sequence [batch_size, seq_len]
+
+    Examples:
+    >>> import torch
+    >>> from x_transformers import TransformerWrapper, Decoder
+    >>> # Create mock transformer
+    >>> B, S, L = 2, 4, 8  # batch_size, seq_len, dim
+    >>> vocab_size = 5
+    >>> model = TransformerWrapper(
+    ...     use_abs_pos_emb=False,
+    ...     num_tokens=vocab_size,
+    ...     max_seq_len=10,
+    ...     attn_layers=Decoder(dim=L, depth=1, heads=2, rotary_pos_emb=True)
+    ... )
+
+    >>> # Test basic generation
+    >>> budget = GenerationBudget.from_seq_len(2)
+    >>> prompts = torch.randint(0, vocab_size, (B, S))
+    >>> gen_tokens = generate(model=model,
+    ...     prompts=prompts,
+    ...     budget=budget,
+    ...     temperature=1.0,
+    ...     filter_logits_fn='top_k',
+    ...     filter_kwargs={'k': 10},
+    ... )
+    >>> gen_tokens.shape  # Should be [batch_size, seq_len]
```

Comment:
These tests are pretty limited, and I don't know if you really need all of them; you could also simplify them by sharing parameters across cases directly to make it more useful as documentation.